lib/logstash/inputs/kafka.rb

```diff
@@ -292,14 +319,24 @@ def thread_runner(logstash_queue, consumer)
                 end
                 event.remove("message")
               end
-              if @decorate_events
+              if @metadata_mode.include?(:record_props)
                 event.set("[@metadata][kafka][topic]", record.topic)
                 event.set("[@metadata][kafka][consumer_group]", @group_id)
                 event.set("[@metadata][kafka][partition]", record.partition)
                 event.set("[@metadata][kafka][offset]", record.offset)
                 event.set("[@metadata][kafka][key]", record.key)
                 event.set("[@metadata][kafka][timestamp]", record.timestamp)
               end
+              if @metadata_mode.include?(:headers)
+                for header in record.headers do
+                  s = String.from_java_bytes(header.value)
+                  begin
+                    s.encode!("UTF-8")
```

Comment:
I'm not 100% familiar with the implementation of JRuby's `String::from_java_bytes`, but `String#encode!` is a pass-through errorless operation when the input is tagged as being in the destination encoding -- whether or not the underlying byte sequence is valid in that encoding.

> ~~~
> irb(main):001:0> shift_jis = "日本".encode('Shift_JIS')
> => "\x{93FA}\x{967B}"
> irb(main):002:0> shift_jis.valid_encoding?
> => true
> irb(main):003:0> shift_jis.force_encoding('UTF-8') # mutates encoding tat but not bytes
> => "\x93\xFA\x96{"
> irb(main):004:0> shift_jis.valid_encoding?
> => false
> irb(main):005:0> shift_jis.encode!('UTF-8')
> => "\x93\xFA\x96{"
> irb(main):006:0> shift_jis.encoding
> => #<Encoding:UTF-8>
> irb(main):007:0> shift_jis.valid_encoding?
> => false
> ~~~

Reply:
So the correct way is to force the encoding in UTF8 and then check if the string is in a validig encoding?
```ruby
s = String.from_java_bytes(header.value)
s.force_encoding(Encoding::UTF-8)
s.valid_encoding?
```